### PR TITLE
Implement PEP 592: Yanked Releases

### PR DESCRIPTION
**Pull request by dstufft**
- [x] Update the ``Release`` model to support marking a release as yanked.
- [x] Update the simple API so that it exposes the ``data-yanked`` attribute.
- [x] ~Provide an API for yanking/unyanking a release.~
- [x] Tests

### CODE REVIEW

**Comment by nlhkabu** [comment]
Note:
If / when this PR is re-reviewed, we will need to ensure that the templates are appropriately translated.

**Comment by brainwane** [comment]
This feature would be helpful in [improving pip's test automation](https://wiki.python.org/psf/PackagingWG/2020-01-23-pip#Automation_snags_and_permissions) so it would be great if you could prioritize finishing this WIP.

**Comment by dstufft** [comment]
I am unlikely to find time in the interim to finish this, though I think we could deploy this without an API and just have it be a UI only feature for now.

**Comment by brainwane** [comment]
@ewdurbin @di @dstufft hey - as Ernest and @pradyunsg  and I [talked about in a meeting about the pip resolver work](https://wiki.python.org/psf/PackagingWG/2020-01-23-pip), getting this implemented might help us smooth the second-order effects of the resolver rollout, so if you could help get this finished and merged, we'd appreciate that!

**Comment by di** [comment]
@nlhkabu, could you take a look at https://github.com/pypa/warehouse/pull/5838/commits/1d7824d969762d09a5222847b8fb7db50c2b0ac8 if you have a second? Without this change, the callout blocks were appearing above the modals.

**Comment by ewdurbin** [comment]
The value of `yanked` be added to the JSON documents for projects/releases, rather than hiding them.

File: warehouse/packaging/models.py
Comment (dstufft) [inline warehouse/packaging/models.py:146]: Do we want to do this? This will make it so yanked versions can't be viewed online at all right?

**Review by dstufft — COMMENTED** [review]
I think my biggest thing is I'm not sure that we want yanked to function as hiding releases from the UI (but maybe we do, I'm not sure?).

My gut instinct tells me we probably want this behavior:

- On the unversioned URL for a project, we won't pick a yanked version, and would instead act as if there were no versions.
  - One could argue that it's better to just display a yanked version as a last ditch effort to display *something*, I'm not sure though?
- On a versioned URL, yanked releases will show up normally, except there is some marker on the page to indict this release has been yanked (likely some sort of red warning banner?).
- On the list of all versions, the yanked releases still show up, but are marked in some fashion to show that they have been yanked (greyed out to demphasize? red? idk).

Fundamentally, I don't think we want to reimplement the hidden release option from legacy PyPI, this is largely about making it so that people can prevent a version from being installed unless absolutely needed, it isn't about hiding it like it never existed and I think that being able to introspect it in the UI still is still useful.

**Comment by di** [comment]
I see where you're coming from @dstufft. My thought is that if we want people to use yanks instead of deleting releases, we need to make yanks behave just like deleting releases, except you can still pin to them. People delete releases for a lot of reasons: bad `long_description`, typos, etc. Ideally in those cases, maintainers can still "remove" a release from the public eye, but anyone already using it can continue to pin to it.

Another issue this helps avoid is requiring average users to understand what a yanked release even is. If a user is looking at project page for `some-project` and they see that the latest release, `1.0.0` is marked as "yanked" with a big red box, but it's still there, they're probably not going to understand what's happening when they `pip install some-project` and get version `0.9.9` instead of `1.0.0`. If the project page just shows the un-yanked releases though, nothing changes here, and they never need to know that `1.0.0` even exists.

**Comment by di** [comment]
Hmm, I see your point, and I think you're probably right. I hadn't considered the difference in behaviors between clients that do/don't support yanked releases.

**Comment by di** [comment]
@dstufft I think I addressed all your feedback here. Yanked releases will show up everywhere except:

* `Project.latest_version` will not return a yanked release as the latest
* The project detail page will not show a yanked release as the latest
* `/pypi/<project_name>/json` will not use a yanked release as the latest
* Search reindexing will not include yanked releases

File: warehouse/templates/legacy/api/simple/detail.html
Comment (dstufft) [inline warehouse/templates/legacy/api/simple/detail.html:22]: This conditional looks weird, why is there two identical conditionals nested inside each other? Why are we setting the ``data-yanked`` attribute equal to the string ``file.release.yanked``?

File: warehouse/templates/pages/help.html
Comment (dstufft) [inline warehouse/templates/pages/help.html:218]: The PEP number is wrong here in the copy.

File: warehouse/templates/legacy/api/simple/detail.html
Comment (di) [reply warehouse/templates/legacy/api/simple/detail.html:22]: Looks like that was a holdover from https://github.com/pypa/warehouse/pull/5838/commits/f976d0abf1156990a5c34db8b8130fcde1796f02, where we might have had a comment or something? I'm not sure what your original intention was, but I updated this just to show the attribute when the release is yanked.

File: warehouse/templates/email/unyanked-project-release/body.html
Comment (pradyunsg) [inline warehouse/templates/email/unyanked-project-release/body.html:19]:
```suggestion
    <li>The {{ project }} release {{ release }} released on {{ release_date }} has been un-yanked.</li>
```

Because the email is for "unyanked-project-release".

File: warehouse/templates/email/unyanked-project-release/body.html
Comment (pradyunsg) [inline warehouse/templates/email/unyanked-project-release/body.html:21]:
```suggestion
      <strong>Un-yanked by:</strong> {{ submitter }} with a role: {{ submitter_role }}.
```

File: warehouse/templates/email/unyanked-project-release/body.txt
Comment (pradyunsg) [inline warehouse/templates/email/unyanked-project-release/body.txt:20]:
```suggestion
  Un-yanked by: {{ submitter }} with a role: {{ submitter_role }}.
```

File: warehouse/templates/email/unyanked-project-release/body.txt
Comment (pradyunsg) [inline warehouse/templates/email/unyanked-project-release/body.txt:18]:
```suggestion
  The {{ project }} release {{ release }} released on {{ release_date }} has been un-yanked.
```

**Review by pradyunsg — COMMENTED** [review]
I feel like it'd be nice to require users to put in something as "reason for yanking" (picking from a bunch of fixed choices, for example) since installers can present that information to the user (pip prints a warning containing the message), and having it be compulsory from the start would help avoid a potential transition step if we add it later.

Then again, it's much easier to do a smoother transition on web application than a command line application, so it's perfectly fine if y'all think it's OK to not start with that... :)

File: warehouse/templates/manage/release.html
Comment (dstufft) [inline warehouse/templates/manage/release.html:142]: This doesn't need to block this PR, but I've wondered if our propensity for linking to PEPs is the best user experience we can provide here, or if we should figure out a real documentation pipeline for Warehouse (not really counting what exists now since it's primarily developer facing documentation, not user facing) and start moving to that.

**Review by dstufft — COMMENTED** [review]
I can't approve this because it's my own PR, but we'll just pretend I did.

File: warehouse/templates/manage/release.html
Comment (di) [reply warehouse/templates/manage/release.html:142]: I think I agree, we are sort of hitting the limits of what we can document inline and in /help.

Comment (brainwane) [reply warehouse/templates/manage/release.html:142]: I filed #7817 - thanks for the idea, @dstufft

**Review by ewdurbin — APPROVED** [review]

File: docs/api-reference/json.rst
Comment (merwok) [inline docs/api-reference/json.rst:231]: Comma missing on the line before (json fun)
Same for line 248

**Comment by pradyunsg** [comment]
YAY! Thanks for all of your work on this @nlhkabu @di @dstufft @ewdurbin! ^>^

**Pull request merged**
